File: src/main/kotlin/com/huawei/cangjie/descriptors/VariableDescriptor.java
package com.huawei.cangjie.descriptors;

import com.huawei.cangjie.types.TypeSubstitutor;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public interface VariableDescriptor extends CallableMemberDescriptor, VariableDescriptorBase {
    @Override
    CallableMemberDescriptor.@NotNull CopyBuilder<? extends VariableDescriptor> newCopyBuilder();

    @NotNull
    @Override
    VariableDescriptor substitute(@NotNull TypeSubstitutor substitutor);

    @Override
    void setOverriddenDescriptors(@NotNull Collection<? extends CallableMemberDescriptor> overriddenDescriptors);

    @Override
    @NotNull
    VariableDescriptor getOriginal();

    @Override
    @NotNull
    Collection<? extends VariableDescriptor> getOverriddenDescriptors();
}

File: src/main/kotlin/com/huawei/cangjie/descriptors/VariableDescriptor.kt
package com.huawei.cangjie.descriptors

import com.huawei.cangjie.resolve.constants.ConstantValue
import com.huawei.cangjie.types.TypeSubstitutor

interface VariableDescriptor : ValueDescriptor ,MemberDescriptor/*,
   CallableMemberDescriptor, VariableSymbolMarker*/ {


    fun getCompileTimeInitializer(): ConstantValue<*>?  = null

    /**
     * ONLY FOR IDE USE! Please don't use the method inside the compiler
     */
    fun cleanCompileTimeInitializerCache()  {}


    /**
     * @return true if iff original declaration has appropriate flags and type, e.g. `const` modifier in CangJie.
     * It completely does not means that if isConst then `getCompileTimeInitializer` is not null
     */
    //    @Nullable
    //    FieldDescriptor getBackingField();
    //
    //    @Nullable
    //    FieldDescriptor getDelegateField();
    override fun substitute(substitutor: TypeSubstitutor): VariableDescriptor?
    val isConst: Boolean

    //    bool isActual();
    //
    //    bool isExternal();
    val isVar: Boolean
}
